### ISSUE TITLE
flag an error when a function cannot return all of its unioned types

### ISSUE THREAD

**Issue by gpshead** [microsoft/pyright#10731]
```python
def beans_api(...) -> Spam | None:
    if failsafe:
        return None

    spam = Spam(...)

    if spam.bad_condition:
        return None   
```

Spot the bug.  A type checker knows that no code path returned Spam despite the type annotation saying it could.

I can imagine cases where this might not be wanted, are they real?  or just worthy of a type checking disable comment.  ex: abstract base class functions... but those are generally in the always-raises NotImplemented category where there is never a return value? a case which should not trigger this. ("never returns" being distinct from "returns X|Y but not Z")

**Comment by erictraut** [comment]
There are many legit reasons for a return type inference to be wider than what a function's or method's current implementation returns. For that reason, such an error would be pretty noisy (lots of false positives). It's definitely not something that could be enabled by default, even in strict mode. I generally don't like to add checks that cannot be enabled by default in strict mode because they tend to get little or no use for discoverability reasons.

I'm also not convinced that this situation, even if it's somewhat legit, leads to real-world bugs going unfound in real code.

For these reasons, I'm going to reject this enhancement request. I'm willing to reconsider this decision in the future if there's sufficient signal from pyright users that this is something they want to see (e.g. in the form of upvotes on this issue).
